openmoc/map_to_dict.i

```diff
@@ -126,13 +126,65 @@
     }
 
     /* Loop over y */
-    for (int j = 0; j < $2; j++) {
+    for (int i =0; i < $2; i++) {
       /* Extract the value from the list at this location and convert
        * SWIG wrapper to pointer to underlying C++ class instance */
-      PyObject* o = PyList_GetItem(outer_list, j);
+      PyObject* o = PyList_GetItem(outer_list, i);
       void *p1 = 0;
       SWIG_ConvertPtr(o, &p1, SWIGTYPE_p_Universe, 0 | 0);
-      $3[i*$2+j] = (Universe*) p1;
+      $3[j*$2+i] = (Universe*) p1;
+    }
+  }
+}
+
+
+/* Typemap for Lattice::setUniverses3D(int num_z, int num_y, int num_x, 
+ *                                     Universe** universes)
+ * method - allows users to pass in Python lists of Universes for each
+ * lattice cell */
+%typemap(in) (int num_z, int num_y, int num_x, Universe** universes) {
+
+  if (!PyList_Check($input)) {
+    PyErr_SetString(PyExc_ValueError,"Expected a Python list of integers "
+                    "for the Lattice cells");
+    return NULL;
+  }
+
+  $1 = PySequence_Length($input);  // num_z
+  $2 = PySequence_Length(PyList_GetItem($input,0)); // num_y
+  $3 = PySequence_Length(PyList_GetItem(PyList_GetItem($input,0), 0)); // num_x
+  $4 = (Universe**) malloc(($1 * $2 * $3) * sizeof(Universe*)); // universes
+
+  /* Loop over y */
+  for (int k = 0; k < $1; k++) {
+
+    /* Get the inner list in the nested list for the lattice */
+    PyObject* outer_outer_list = PyList_GetItem($input,k);
+    
+    /* Loop over y */
+    for (int j = 0; j < $2; j++) {
+
+      /* Get the inner list in the nested list for the lattice */
+      PyObject* outer_list = PyList_GetItem(outer_outer_list, j);
+
+      /* Check that the length of this list is the same as the length
+       * of the first list */
```

Comment:
I agree, sounds good

On Mon, Oct 12, 2015 at 6:53 PM, sam <notifications@github.com> wrote:

> In openmoc/map_to_dict.i
> <https://github.com/mit-crpg/OpenMOC/pull/188#discussion_r41809937>:
>
> > +  $4 = (Universe**) malloc(($1 * $2 * $3) * sizeof(Universe*)); // universes
> > +
> > +  /* Loop over y */
> > +  for (int k = 0; k < $1; k++) {
> > +
> > +    /* Get the inner list in the nested list for the lattice */
> > +    PyObject* outer_outer_list = PyList_GetItem($input,k);
> > +
> > +    /* Loop over y */
> > +    for (int j = 0; j < $2; j++) {
> > +
> > +      /* Get the inner list in the nested list for the lattice */
> > +      PyObject* outer_list = PyList_GetItem(outer_outer_list, j);
> > +
> > +      /* Check that the length of this list is the same as the length
> > +       * of the first list */
>
> Also, I think something like "Check that the number of universes in the
> j-th row of the k-th xy-plane is the same as the number of cells in the 1st
> row of the 1st z-plane" would more accurately describe what this is check
> for, physically.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/188/files#r41809937>.
>



-- 
Will Boyd
Nuclear Science & Engineering
Massachusetts Institute of Technology
Cell: 423.413.8469
wbinventor@gmail.com